theatre/studio/src/panels/SequenceEditorPanel/DopeSheet/Right/BasicKeyframedTrack/KeyframeEditor/CurveEditorPopover/CurveSegmentEditor.tsx

```diff
@@ -0,0 +1,288 @@
+
+      <path
+        d={`M0 ${toExtremumSpace(1)} C${cur.handles[2]} ${toExtremumSpace(
+          1 - cur.handles[3],
+        )} 
+    ${next.handles[0]} ${toExtremumSpace(
+          1 - next.handles[1],
+        )} 1 ${toExtremumSpace(0)}`}
+        stroke="none"
+        fill="url('#myGradient')"
+        opacity="0.1"
+      />
+      <path
+        d={`M0 ${toExtremumSpace(1)} C${cur.handles[2]} ${toExtremumSpace(
+          1 - cur.handles[3],
+        )} 
+    ${next.handles[0]} ${toExtremumSpace(
+          1 - next.handles[1],
+        )} 1 ${toExtremumSpace(0)}`}
+        stroke="url('#myGradient')"
+        strokeWidth="0.02"
+      />
+
+      <circle
+        cx={0}
+        cy={toExtremumSpace(1)}
+        r="0.04"
+        stroke={CURVE_START_COLOR}
+        strokeWidth="0.02"
+        fill="black"
+      />
+
+      <circle
+        cx={1}
+        cy={toExtremumSpace(0)}
+        r="0.04"
+        stroke={CURVE_END_COLOR}
+        strokeWidth="0.02"
+        fill="black"
+      />
+
+      <HitZone
+        ref={refLeft}
+        cx={cur.handles[2]}
+        cy={toExtremumSpace(1 - cur.handles[3])}
+        fill={CURVE_START_COLOR}
+        opacity={0.2}
+      />
+      <Circle
+        cx={cur.handles[2]}
+        cy={toExtremumSpace(1 - cur.handles[3])}
+        fill={CURVE_START_COLOR}
+      />
+
+      <HitZone
+        ref={refRight}
+        cx={next.handles[0]}
+        cy={toExtremumSpace(1 - next.handles[1])}
+        fill={CURVE_END_COLOR}
+        opacity={0.2}
+      />
+      <Circle
+        cx={next.handles[0]}
+        cy={toExtremumSpace(1 - next.handles[1])}
+        fill={CURVE_END_COLOR}
+      />
+    </svg>
+  )
+}
+export default CurveSegmentEditor
+
+function useKeyframeDrag(
+  svgNode: SVGSVGElement | null,
+  node: SVGCircleElement | null,
+  props: IProps,
+  setKeyframe: (dx: number, dy: number) => Keyframe,
+): void {
+  const handlers = useFreezableMemo<Parameters<typeof useDrag>[1]>(
+    (setFrozen) => {
+      // Considered using "scrub" instead to manage a not-necessarilly commital change over time
+      // But it appears that scrub doesn't allow to change a "pointer" for the keyframes
+      let tempTransaction: CommitOrDiscard | undefined
+
+      return {
+        debugName: 'CurveSegmentEditor/useRightDrag',
+        lockCursorTo: 'move',
+        onDragStart() {
+          setFrozen(true)
+        },
+        onDrag(dx, dy) {
+          if (!svgNode) return
+          tempTransaction?.discard()
+          tempTransaction = undefined
+
+          tempTransaction = getStudio()!.tempTransaction(({stateEditors}) => {
+            stateEditors.coreByProject.historic.sheetsById.sequence.replaceKeyframes(
+              {
+                ...props.leaf.sheetObject.address,
+                snappingFunction: val(props.layoutP.sheet).getSequence()
+                  .closestGridPosition,
+                trackId: props.leaf.trackId,
+                keyframes: [setKeyframe(dx, dy)],
```

Comment:
This is very tricky. I tried and failed. Background: our dragged position is currently calculated using the control handle positions at the start of the drag and the drag dx and dy. Here is the approach I took:

**Scaling the vertical drag amount by the current height of the unit space.** This fails because the curve height scales the vertical drag, which in turn scales the curve height, and on and on in a loop. It results in compounding and massive vertical drag movements when dragging out of the [0,1] vertical space. AND even when dragging in to the space, the cursor gets offset from the handle as the extremum space changes. This approach does fix it so that once you get back to `extremum space = unit space` then there is no drift, but at the expense that your cursor is now offsets.

By the way I am currently thinking about this, in order to mitigate the compounding effect and mitigate the cursor drift as the extremum space is changing, I think you would either need to do some calculus or think about the dragging in a different way.

Reply:
Okay I think I missed that the drift only occurs with overshoots/space-bending, so I think it's fine. I'll mark this as resolved. Thanks for the explanation.